[1.20.4/5] Create BonusLootDropsEvent

AnonymousHacker1279 opened the pull request:
Create a `BonusLootDropsEvent` which is fired before dropping items via the `ApplyBonusCount` loot function (for example, blocks with Fortune modifiers). It allows enchantment levels to be modified or a specific drop count to be set.

All of the variables/parameters available in `ApplyBonusCount.run()` are exposed through the event.

Example Usages:
```java
@SubscribeEvent
public static void bonusLootDropsEvent(BonusLootDropsEvent event) {
	// Add two extra enchantment levels when specific conditions are met
	if (event.getEntity() instanceof Player player) {
		if (playerMeetsSpecificConditions()) {
			event.setEnchantmentLevel(event.getEnchantmentLevel() + 2);
		}
	}
}
```
```java
@SubscribeEvent
public static void bonusLootDropsEvent(BonusLootDropsEvent event) {
	// Always set the drops to three when specific conditions are met
	if (event.getEntity() instanceof Player player) {
		if (playerMeetsSpecificConditions()) {
			event.setDropCount(3);
		}
	}
}
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/neoforged/NeoForge?pullRequest=625) <br/>All committers have signed the CLA.

Shadows-of-Fire commented on patches/net/minecraft/world/level/storage/loot/functions/ApplyBonusCount.java.patch:
```diff
@@ -0,0 +1,28 @@
+     }
+ 
+-    interface Formula {
++    public interface Formula {
```
should be an AT

AnonymousHacker1279 replied:
AT added in latest commit

Shadows-of-Fire commented on patches/net/minecraft/world/level/storage/loot/functions/ApplyBonusCount.java.patch:
```diff
@@ -0,0 +1,28 @@
+     }
+ 
+-    static record FormulaType(ResourceLocation id, Codec<? extends ApplyBonusCount.Formula> codec) {
++    public static record FormulaType(ResourceLocation id, Codec<? extends ApplyBonusCount.Formula> codec) {
```
should be an AT

AnonymousHacker1279 replied:
AT added in latest commit

Shadows-of-Fire commented:
Not 100% convinced this is a great solution, especially because changing the enchantment level is totally superseded by someone else modifying the drop count.

AnonymousHacker1279 commented:
The drop count method intentionally acts as a final say-so in case a mod decides it needs to force it for any reason. I can change this if desired though.

marchermans commented on build.gradle:
```diff
@@ -22,7 +22,7 @@ gradleutils.version {
 }
 
 changelog {
-    from '20.4'
```
No, obviously this was not checked!

marchermans commented on patches/com/mojang/blaze3d/vertex/VertexFormatElement.java.patch:
```diff
@@ -11,23 +11,3 @@
-     }
- 
-     @OnlyIn(Dist.CLIENT)
--    public static enum Usage {
-+    public static enum Usage implements net.neoforged.neoforge.common.IExtensibleEnum {
-         POSITION("Position", (p_167043_, p_167044_, p_167045_, p_167046_, p_167047_, p_167048_) -> {
-             GlStateManager._enableVertexAttribArray(p_167048_);
-             GlStateManager._vertexAttribPointer(p_167048_, p_167043_, p_167044_, false, p_167045_, p_167046_);
-@@ -188,6 +_,10 @@
-         @OnlyIn(Dist.CLIENT)
-         public interface SetupState {
-             void setupBufferState(int p_167053_, int p_167054_, int p_167055_, long p_167056_, int p_167057_, int p_167058_);
-+        }
-+
-+        public static Usage create(String name, String usageName, SetupState setupState, ClearState clearState) {
-+            throw new IllegalArgumentException("Enum not extended");
-         }
-     }
- }
```
Your merge completely failed!

marchermans reviewed (changes requested):
I am not going to review this further.
This PR has had a merge failure at some point and I suggest starting a new with a fresh branch and then force pushing to your old branch to get it back into a working state.

Turning this PR into a draft.

AnonymousHacker1279 committed:
Create BonusLootDropsEvent

Create a BonusLootDropsEvent which is fired before dropping items via the ApplyBonusCount loot function. It allows enchantment levels to be modified or a specific drop count to be set.

AnonymousHacker1279 commented:
My apologies, I did not catch the merge failure when updating my branch through GitHub Desktop. I have force-pushed this branch from a fresh one.

marchermans commented:
Okey that merge disaster is indeed fixed, now to the content.

Would a GLM not work here?

AnonymousHacker1279 commented:
The problems with GLMs for this particular use scenario is that they run *after* loot is generated, and Fortune-style modifiers run before.

There was a [more extended discussion](https://discord.com/channels/313125603924639766/852298000042164244/1205968935498022962) on Discord regarding this.

KnightMiner commented:
In the past, the forge recommendation for how to deal with this issue is to use a GLM to delete all generated loot then create a modified context to rerun the loot table. Obviously this is a horrible solution as it means other GLMs trying a similar approach are clobbered, but this is why this has never been addressed.

Not sure if the silk touch GLM doing that is still around, that case at least is long since handled by item enchantment overrides.

----

This PR to me seems like it has notable overlap #503 which was recently merged. It really needs a javadoc on both events explaining when this should be used over that and vice versa.

The only usecase for the enchantment side of this event I can think of is if you need to set enchantments conditioned on the details of the loot context (e.g. you increase fortune conditioned on the block being mined) as the enchantment event is better for other cases as it lets non-loot table contexts properly query the enchantment. I'd normally say you should add an example of why you need that functionality, but in this case its functionality I use in one of my mods (fortune conditioned on only a subset of blocks that can be mined)

AnonymousHacker1279 commented:
I was referred to the `GetEnchantmentLevelEvent` introduced in #503 however my particular use case was being able to adjust Fortune levels based on an armor set bonus, meaning I needed a player context of some sort. It was mentioned in Discord that ideally the `GetEnchantmentLevelEvent` would contain this context, but the patch surface was too high to be feasible. 

Later this evening I can improve the JavaDocs on this event to explain when it should be used over the other event.

AnonymousHacker1279 committed:
Update event JavaDoc

Improve the event JavaDoc with additional information and a recommendation to use the GetEnchantmentLevelEvent when possible.

sciwhiz12 commented:
Hello! Do you intend to continue with this PR? If so, please fix the merge conflicts.

AnonymousHacker1279 commented:
I have no intentions of continuing the PR at this point.
